src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,831 @@
+    fn connector_type(&self) -> ConnectorType {
+        ConnectorType("gcs_writer".into())
+    }
+
+    async fn build_cfg(
+        &self,
+        _alias: &str,
+        _config: &ConnectorConfig,
+        connector_config: &Value,
+        _kill_switch: &KillSwitch,
+    ) -> Result<Box<dyn Connector>> {
+        let config = Config::new(connector_config)?;
+
+        if config.buffer_size % (256 * 1024) != 0 {
+            return Err("Buffer size must be a multiple of 256kiB".into());
+        }
+        let url = Url::<HttpsDefaults>::parse(&config.endpoint)?;
+
+        Ok(Box::new(GCSWriterConnector { config, url }))
+    }
+}
+
+struct GCSWriterConnector {
+    config: Config,
+    url: Url<HttpsDefaults>,
+}
+
+#[async_trait::async_trait]
+impl Connector for GCSWriterConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let default_bucket = self
+            .config
+            .bucket
+            .as_ref()
+            .map(|bucket_name| Value::from(bucket_name.as_str()).into_static());
+        let sink = GCSWriterSink {
+            client_tx: None,
+            url: self.url.clone(),
+            config: self.config.clone(),
+            buffers: ChunkedBuffer::new(self.config.buffer_size),
+            current_name: None,
+            default_bucket,
+            done_until: Arc::new(AtomicUsize::new(0)),
+        };
+
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
+    data: Vec<u8>,
+    block_size: usize,
+    buffer_start: usize,
+}
+
+impl ChunkedBuffer {
+    pub fn new(size: usize) -> Self {
+        Self {
+            data: Vec::with_capacity(size * 2),
+            block_size: size,
+            buffer_start: 0,
+        }
+    }
+
+    pub fn mark_done_until(&mut self, position: usize) -> Result<()> {
+        if position < self.buffer_start {
+            return Err("Buffer was marked as done at index which is not in memory anymore".into());
+        }
+
+        let bytes_to_remove = position - self.buffer_start;
+        self.data = Vec::from(self.data.get(bytes_to_remove..).ok_or(
+            ErrorKind::GoogleCloudStorageError("Not enough data in the buffer"),
+        )?);
+        self.buffer_start += bytes_to_remove;
+
+        Ok(())
+    }
+
+    pub fn read_current_block(&self) -> Option<&[u8]> {
+        self.data.get(..self.block_size)
+    }
+
+    pub fn write(&mut self, data: &[u8]) {
+        self.data.extend_from_slice(data);
+    }
+
+    pub fn start(&self) -> usize {
+        self.buffer_start
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn final_block(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+enum HttpTaskCommand {
+    FinishUpload {
+        name: String,
```

Comment:
We use `bucket` in `StartUpload` so could this collide with having files with the same name but different buckets?

Reply:
looking at the code below, yap that would clash